File: SquidstatLibrary/examples/Python/firmwareUpdate.py
"""! @example firmwareUpdate.py """
##! [FirmwareUpdate]
import sys
from PySide6.QtCore import QCoreApplication

from SquidstatPyLibrary import AisDeviceTracker
from SquidstatPyLibrary import AisInstrumentHandler
from SquidstatPyLibrary import AisErrorCode

# Define relavant device information, for easy access
COMPORT = "COM16"

app = QCoreApplication()

tracker = AisDeviceTracker.Instance()

def onProgressMessage(message):
    print(message)
    if message.__contains__("firmware is updated"):
        app.quit()

##! [FirmwareUpdateNotification]
tracker.firmwareUpdateNotification.connect(onProgressMessage)
##! [FirmwareUpdateNotification]

# Attempt to connect to the device
error = tracker.connectToDeviceOnComPort(COMPORT)
if error.value() == AisErrorCode.FirmwareNotSupported:
    error = tracker.updateFirmwareOnComPort(COMPORT)
    
    # Some other error occured
    if error.value() != AisErrorCode.Success:
        print(f"Error: {error.message()}")
        sys.exit()
elif error.value() != AisErrorCode.Success:
    print(f"Error: {error.message()}")
    sys.exit()
else:
    print("Device is already up to date.")
    sys.exit()
    
sys.exit(app.exec())
##! [FirmwareUpdate]

File: SquidstatLibrary/examples/C++/firmwareUpdate/firmwareUpdate.cpp
/** \example firmwareUpdate.cpp */
//! [FirmwareUpdate]
#include "AisInstrumentHandler.h"
#include "AisDeviceTracker.h"

#include <QCoreApplication>
#include <QThread>
#include <QDebug>

#define COMPORT "COM1"

int main()
{
    int args;
    QCoreApplication a(args, nullptr);

    auto tracker = AisDeviceTracker::Instance();

    //! [FirmwareUpdateNotification]
    QObject::connect(tracker, &AisDeviceTracker::firmwareUpdateNotification, [=](const QString& message) {
        qDebug() << message;
        if (message.contains("firmware is updated.")) {
            // Now instrument is ready to go
        }
    });
    //! [FirmwareUpdateNotification]

    // Attempt to connect to the device
    auto error = tracker->connectToDeviceOnComPort(COMPORT);
    if (error == AisErrorCode::FirmwareNotSupported) {
        error = tracker->updateFirmwareOnComPort(COMPORT);

        // Some other error occured
        if (error != AisErrorCode::Success) {
			qDebug() << "Error: " << error.message();
		}

    } else if (error != AisErrorCode::Success) {
        qDebug() << "Error: " << error.message();
    } else {
        qDebug() << "Device firmware is up to date";
    }

    return a.exec();
}
//! [FirmwareUpdate]